[1단계 - 보드 초기화, 기물 이동] 비타(오연택) 미션 제출합니다.

taek2222 opened the pull request:
## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [ ] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.

미션을 수행하면서 의식적으로 지키기 위해 노력했고, 추후 리팩토링 시간을 충분히 확보해 페어와 함께 이를 지키기 위해 다듬었습니다! ☺️
  
## 어떤 부분에 집중하여 리뷰해야 할까요?

### 더 좋은 설계

>📌 미션을 수행하며 반드시 고민할 것
> ✅ 어떤 원칙을 적용해야 더 좋은 설계를 할 수 있을까?

페어와 함께 `더 좋은 설계`라는 키워드를 바탕으로 충분히 고민하며 작성했습니다.

다만, 현재 저희 코드에서 ‘더 좋은 설계’라는 키워드에 부합하는 명확한 방향성을 찾는 데 어려움을 느껴 더 나아가지 못했습니다. 😢

개선 가능하거나 놓친 부분을 짚어주시면 페어와 함께 다시 깊게 고민하고 반영해 보겠습니다!

---

### 장기 기물 초기화(위치, 방향)
장기 기물을 초기화하는 과정에서 `PieceDirection`과 `PieceInit` 클래스를 사용하고 있습니다.

`Enum`의 주된 목적이 상수 정의라고 알고 있는데, 장기의 초기 기물 배치가 고정된 `상수`라는 점을 고려하여 `Enum` 내부에서 직접 객체를 생성하고 이를 `상수`로 관리하는 방식을 선택했습니다. 🤔

리뷰어님의 관점에서는 이런 접근이 적절하다고 생각하시는지 궁금합니다. 👀

---

### 리뷰어께 ✍️
**`LEVEL 1`** 과정 동안 **자바**와 **객체지향**의 개념을 확실하게 다잡겠다고 다짐했습니다. 🔥

마지막 미션인 장기를 진행하며 부족한 역량을 충분히 학습하고 체득하여, 보다 의식적으로 객체지향적인 사고를 할 수 있는 개발자로 성장하고 싶습니다. 

이를 위해 리뷰어님의 피드백을 최대한 많이 받고 적극적으로 개선하겠습니다. 잘 부탁드립니다! 🙇🏻

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **문서**
  - 게임 설명서가 전면 개편되어, 장기의 기능 목록, 기물 설명, 게임 시작 안내 및 실행 예시가 추가되었습니다.

- **새로운 기능**
  - 장기 게임의 전반적인 플레이 흐름(보드 설정, 기물 이동, 에러 처리, 승리 조건 등)이 새롭게 구현되었습니다.
  - 사용자 입력 및 출력 인터페이스가 강화되어, 보다 명확한 프롬프트와 보드 표기로 게임 경험이 향상되었습니다.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

taek2222 committed:
docs: 기능 요구 사항 초안 작성

taek2222 committed:
docs: 기능 요구 사항 초안 수정

rladmstn committed:
test: 좌표 생성 테스트 작성

rladmstn committed:
feat: 좌표 생성 구현

rladmstn committed:
docs: 예외 처리 추가

taek2222 committed:
test: 잘못된 입력인 경우 테스트 추가

taek2222 committed:
feat: 잘못된 입력 검증 추가

taek2222 committed:
test: 잘못된 입력 검증 파라미터 추가

taek2222 committed:
refactor: 잘못된 좌표 입력 검증 개선

rladmstn committed:
test: 기물 객체 생성 테스트 작성

rladmstn committed:
feat: 기물 객체 생성 기능 구현

taek2222 committed:
chore: 기능 구현 리드미 위치 변경

taek2222 committed:
feat: 장기 기물 종류 추가

taek2222 committed:
docs: 변경된 출력 수정 및 요구 사항 체크

taek2222 committed:
feat: 장기 기물 이름 추가

taek2222 committed:
docs: 장기 기물 이름 변경

rladmstn committed:
test: 기물 리스트 반환 테스트 작성

rladmstn committed:
feat: 기물 리스트 반환 기능 구현

taek2222 committed:
feat: 장기 초기 기물 좌표 입력

taek2222 committed:
test: 장기판 초기화 테스트 코드 추가

taek2222 committed:
feat: 보드 초기화 기능 추가

rladmstn committed:
feat: 기물 리스트 반환 기능 구현

rladmstn committed:
feat: getter 추가

rladmstn committed:
feat: 장기 초기화 출력 비즈니스 로직 추가

rladmstn committed:
feat: 출력 컬러 추가

rladmstn committed:
test: 해당 위치에 도달할 수 있는지 판단하는 테스트 작성

rladmstn committed:
feat: 정적 팩토리 메서드 적용

rladmstn committed:
test: 좌표 병합 테스트 작성

rladmstn committed:
feat: 좌표 병합 기능 구현

rladmstn committed:
feat: 도달 가능한지 판단하는 기능 추가

rladmstn committed:
test: 도달 가능한지 판단하는 파라미터 테스트 추가

taek2222 committed:
test: 이동 경로 반환 테스트 작성

taek2222 committed:
feat: 이동 거리 반환 기능 추가

taek2222 committed:
test: 반복 이동이 있는 방향성 경로 반환 테스트 추가

taek2222 committed:
feat: 반복인 경우 경로 반환 로직 추가

taek2222 committed:
test: Direction 객체 구조 변경으로 요소 변경

rladmstn committed:
test: 좌표 유효성 판단 테스트 추가

rladmstn committed:
feat: 좌표 유효성 판단 기능 추가

rladmstn committed:
test: 반복인 경우 타겟 위치 도달 여부 판단 테스트 작성

rladmstn committed:
feat: 반복인 경우 타겟 위치 도달 여부 판단 기능 추가

rladmstn committed:
fix: 객체 생성 구조 변경

merge인 경우 검증이 불필요한 상황을 개선하기 위해 기본 생성자로 Position을 생성합니다.

taek2222 committed:
test: 타겟 위치 방향을 찾아 이동 경로 반환 테스트 작성

taek2222 committed:
feat: 방향을 찾아 경로를 반환하는 기능을 추가

taek2222 committed:
test: 반복인 경우 방향 올바르게 찾는 테스트 검증 추가

rladmstn committed:
test: 특정 위치에 있는 기물 반환 테스트 작성

rladmstn committed:
feat: 특정 위치에 있는 기물 반환 기능 구현

rladmstn committed:
test: 기물이 타겟 위치까지 도달하는 이동 경로 반환 테스트 작성

rladmstn committed:
feat: 기물 객체에 방향 필드 추가

rladmstn committed:
feat: 방향 초기화 추가

rladmstn committed:
test: 좌표 목록에 일치하는 기물 개수 계산 테스트 작성

rladmstn committed:
test: 위치가 같은지 판단하는 기능 테스트 작성

rladmstn committed:
feat: 위치가 같은지 판단하는 기능 추가

rladmstn committed:
feat: 좌표 목록에 일치하는 기물 계산 기능 구현

taek2222 committed:
Merge remote-tracking branch 'origin/step1' into step1

taek2222 committed:
test: 기물의 좌표를 변경 테스트 작성

taek2222 committed:
test: 기물 위치 변경 테스트 작성

taek2222 committed:
feat: 부 생성자 추가 및 위치 변경 기능 추가

taek2222 committed:
test: 테스트 기물도 변경점 적용

taek2222 committed:
feat: 객체 equals 재정의 추가

taek2222 committed:
feat: 기물 변경 기능 추가

taek2222 committed:
test: 기물 좌표 변경 테스트 불변 제거

taek2222 committed:
feat: 움직이는 기능 추가 기능 개발 단계 정의

rladmstn committed:
test: 플레이어 기물 이동 테스트 작성

rladmstn committed:
feat: 기물 이동 기능 구현

taek2222 committed:
test: 좌표에 기물이 존재하는지 판단 테스트 작성

taek2222 committed:
feat: 좌표에 기물이 존재하는 지 판단 기능 구현

taek2222 committed:
test: 타겟 위치의 아군 기물이 있는 경우 예외 발생 테스트 작성

taek2222 committed:
feat: 타겟 위치의 아군 기물이 있는 경우 예외 발생 추가

taek2222 committed:
test: 다른 포지션 예시로 변경

taek2222 committed:
fix: 이동거리 반환 오류 수정

rladmstn committed:
test: 이동한 위치에 존재하는 상대 기물 삭제 테스트 작성

rladmstn committed:
test: 좌표의 기물 삭제 테스트 작성

rladmstn committed:
feat: 좌표의 기물 삭제 기능 구현

rladmstn committed:
feat: 상대편 기물 삭제 기능 추가

rladmstn committed:
docs: 요구사항 수정

rladmstn committed:
feat: 기물 이동 비즈니스 로직 추가

rladmstn committed:
fix: 졸 방향 오류 수정

taek2222 committed:
test: 게임 종료 여부 테스트 작성

taek2222 committed:
test: 왕이 있는지 판단 테스트 작성

taek2222 committed:
test: 왕이 판단 테스트 작성

taek2222 committed:
feat: 왕 판단 기능 추가

taek2222 committed:
feat: 기물중 왕 존재 판단 기능 추가

taek2222 committed:
test: 왕 존재 판단 경우의 수 추가

taek2222 committed:
feat: 게임 종료 판단 기능 구현

rladmstn committed:
test: 우승자 반환 테스트 작성

rladmstn committed:
feat: 우승자 반환 기능 추가

rladmstn committed:
feat: 우승자 비즈니스 로직 추가

taek2222 committed:
test: 포인지 판단 테스트 작성

taek2222 committed:
feat: 포인지 판단 기능 추가

taek2222 committed:
test: 포 이동 경로 기물 갯수 검증 테스트 작성

taek2222 committed:
feat: 포 이동 경로 기물 갯수 검증 추가

rladmstn committed:
test: 좌포 기물이 포인지 판단 테스트 작성

rladmstn committed:
feat: 좌포 기물이 포인지 판단 기능 구현

rladmstn committed:
test: 포의 목적지에 상대 포가 존재할 경우 예외 발생 테스트 작성

rladmstn committed:
feat: 포의 목적지에 상대 포가 존재할 경우 예외 추가

rladmstn committed:
test: 포의 경로에 다른 포가 존재할 경우 예외 발생 테스트 작성

rladmstn committed:
feat: 포의 경로에 다른 포가 존재할 경우 예외 추가

rladmstn committed:
feat: 최종 출력 포맷팅 정리

rladmstn committed:
refactor: 패키지 분리

rladmstn committed:
refactor: 누락된 final 키워드 추가 및 메서드 재정의

rladmstn committed:
refactor: 가독성 향상을 위한 메서드 분리 및 네이밍 수정

rladmstn committed:
test: 기물 종류 판단 테스트 케이스 추가

rladmstn committed:
docs: 요구사항 구현 기능 체크

hyeonic commented on src/main/java/domain/piece/category/Pawn.java:
```diff
@@ -0,0 +1,37 @@
+import domain.piece.Piece;
+import domain.piece.Position;
+
+public class Pawn extends Piece {
```
체스의 용어를 활용하신 이유가 있을까요? 위키피디아의 장기를 살펴보면 관련 표현은 찾기 어려워서요.

taek2222 replied:
위키피디아가 아닌, [나무위키](https://namu.wiki/w/%EC%9E%A5%EA%B8%B0/%EC%9A%A9%EC%96%B4) 사이트를 참고했습니다.. 🤔

하지만, 이는 **`장기`** 미션에서 주어진 기획서에 명칭이 있었음에도 불구하고, 다른 자료의 명칭을 가져와 사용한 건 잘못된 방향이라 생각합니다. 수정하도록 하겠습니다! 🔥

<div align="center">
  <img src="https://github.com/user-attachments/assets/41fbcbd2-46ab-4ff6-9395-6e7819c7251d" width="30%">
</div>

**변경 코멘트 : [refactor: 명확한 기물 클래스 네이밍 변경](https://github.com/woowacourse/java-janggi/pull/43/commits/ddd5d016f4ac6422bf84b719bfcfef9520f81baa)**

hyeonic commented on src/main/java/domain/piece/Piece.java:
```diff
@@ -0,0 +1,59 @@
+
+    public abstract Piece updatePosition(final Position position);
+
+    public abstract boolean isKing();
+
+    public abstract boolean isCannon();
```
추상 클래스 이지만 상속 받게되는 객체 이름을 강하게 의존하는 것 처럼 보이기도 합니다. Advisor, Elephant 등 다른 기물에서 관련 키워드를 의존해도 괜찮을까요?

taek2222 replied:
말씀해주신 부분에 깊이 공감했습니다. 부모 클래스가 자식 클래스의 특정 키워드를 직접 알고 있다는 점이 저도 어색하게 느껴졌습니다. 🧐

그래서 다른 방안을 고민하다가, 메서드를 삭제하고 instanceof를 활용하는 방향으로 개선할 수도 있겠다고 생각했습니다. 하지만 instanceof 키워드에 대한 자료를 찾아보다가 아래의 참고자료를 발견했습니다. 

[참고자료](https://ksh-coding.tistory.com/84)

해당 자료에서 지적하는 것처럼, instanceof를 사용하는 방식은 캡슐화와 추상화의 원칙을 위반할 우려가 있었습니다. 🧐

현재 방법과 instanceof의 문제점을 서로 비교해보니, 오히려 현재 방식이 비용적인 측면에서 더 나은 선택이라는 판단이 들었습니다. 혹시 이 부분에 대해 매트는 어떻게 생각하시는지 의견이 궁금합니다..! 🤔

hyeonic replied:
instanceof의 사용은 말씀 주신 것 처럼 캡슐화 및 추상화의 원칙을 위반할 수 있을 것 같아서 사용을 지양하는 편이 좋을 것 같아요!

먼저 비타가 설계하신 보드(현재 방식)에서는 모든 피스에 대한 이동 전략에 대해 알고 있기 때문에 이를 판단하기 위한 코드(isKing, isCannon)가 필요할 것 같습니다.

hyeonic commented on src/main/java/domain/piece/PieceInit.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public enum PieceInit {
+
+    KING(new King(5, 2, PieceDirection.KING.get()), new King(5, 9, PieceDirection.KING.get())),
+    ROOK1(new Rook(1, 1, ROOK.get()), new Rook(1, 10, ROOK.get())),
+    ROOK2(new Rook(9, 1, ROOK.get()), new Rook(9, 10, ROOK.get())),
+    CANNON1(new Cannon(2, 3, CANNON.get()), new Cannon(2, 8, CANNON.get())),
+    CANNON2(new Cannon(8, 3, CANNON.get()), new Cannon(8, 8, CANNON.get())),
+    HORSE1(new Horse(3, 1, HORSE.get()), new Horse(3, 10, HORSE.get())),
+    HORSE2(new Horse(7, 1, HORSE.get()), new Horse(7, 10, HORSE.get())),
+    ELEPHANT1(new Elephant(2, 1, ELEPHANT.get()), new Elephant(2, 10, ELEPHANT.get())),
+    ELEPHANT2(new Elephant(8, 1, ELEPHANT.get()), new Elephant(8, 10, ELEPHANT.get())),
+    ADVISOR1(new Advisor(4, 1, ADVISOR.get()), new Advisor(4, 10, ADVISOR.get())),
+    ADVISOR2(new Advisor(6, 1, ADVISOR.get()), new Advisor(6, 10, ADVISOR.get())),
+    PAWN1(new Pawn(1, 4, HAN_PAWN.get()), new Pawn(1, 7, CHO_PAWN.get())),
+    PAWN2(new Pawn(3, 4, HAN_PAWN.get()), new Pawn(3, 7, CHO_PAWN.get())),
+    PAWN3(new Pawn(5, 4, HAN_PAWN.get()), new Pawn(5, 7, CHO_PAWN.get())),
+    PAWN4(new Pawn(7, 4, HAN_PAWN.get()), new Pawn(7, 7, CHO_PAWN.get())),
+    PAWN5(new Pawn(9, 4, HAN_PAWN.get()), new Pawn(9, 7, CHO_PAWN.get())),
+    ;
```
> 장기 기물을 초기화하는 과정에서 PieceDirection과 PieceInit 클래스를 사용하고 있습니다.
> 
> Enum의 주된 목적이 상수 정의라고 알고 있는데, 장기의 초기 기물 배치가 고정된 상수라는 점을 고려하여 Enum 내부에서 직접 객체를 생성하고 이를 상수로 관리하는 방식을 선택했습니다. 🤔
> 
> 리뷰어님의 관점에서는 이런 접근이 적절하다고 생각하시는지 궁금합니다. 👀

enum 활용에 대해서는 충분히 고려할 수 있다고 생각해요. 다만 경우에 따라 장기는 아래와 같이 다양한 셋업 방법을 제공하고 있는 것으로 알고 있는데요

![image](https://github.com/user-attachments/assets/ef2fe748-112a-4a12-bf06-c87703e7bba3)

이럴 때 비타의 enum 활용은 어떠한 단점을 가질 수 있을까요?

taek2222 replied:
위와 같은 기능을 추가할 때, `Enum`을 활용하는 방식의 단점은 확장성이 제한될 수 있다는 점입니다. 이 부분을 해결하기 위해 여러 방안을 고민하다가.. 늦었습니다 😢 최종적으로는, 전략 패턴과 팩토리 패턴을 조합하는 방식이 확장성 측면에서 더 유연한 설계가 될 것이라 판단했습니다.

해당 방식으로 리팩토링을 진행했지만, 일부 중복 코드가 존재하여 추가적인 개선이 필요하다고 느꼈습니다. 🤔
현재 중복되는 부분을 줄이고 더욱 효율적인 구조로 개선해 보려고 하는데, 전략과 팩토리 설계의 방향이 적절할지 의견 부탁드립니다! ☺️

**변경 코멘트 :**

- [docs: 초기 전략 선택 기능 요구 사항 추가](https://github.com/woowacourse/java-janggi/pull/43/commits/e77f8a1935ff9ff850ff085649ffc5785a535df4)

- [feat: 초기 전략 선택 기능 추가](https://github.com/woowacourse/java-janggi/pull/43/commits/4fffef29d1b8671100efb18ef7f07d7499805d73)

- [fix: 잘못된 전략 순서 변경 완료](https://github.com/woowacourse/java-janggi/pull/43/commits/5b0b0d00b67cf9c7d3da15cc77a899cb5e8489d4)

hyeonic commented on src/main/java/domain/piece/category/Cannon.java:
```diff
@@ -0,0 +1,37 @@
+import domain.piece.Piece;
+import domain.piece.Position;
+
+public class Cannon extends Piece {
+
+    private static final String NAME = "c";
+
+    public Cannon(final int row, final int column, final Directions directions) {
+        super(row, column, directions);
+    }
```
비타는 어떤 때에 클래스의 생성자를 통한 주입과 필드를 고려하시나요? Directions을 외부에서 받았을 때 어떠한 이점을 누릴 수 있다고 생각하시나요?

taek2222 replied:
`Directions` 객체를 외부로부터 생성자를 통해 주입받으면, 테스트 상황에서 해당 객체를 `Mock`으로 대체하거나 임의의 방향 설정을 통해 특정 로직을 테스트할 수 있습니다. 이는 유연하고 유지보수하기 쉬운 코드를 작성할 수 있는 장점이 있습니다. 📚

`Position`의 경우에도 외부에서 주입받는 방식을 채택하면 유사한 이점을 얻을 수 있습니다. 그러나 초기 설계 시에는 `Position` 클래스가 단순한 `VO 객체`로서 좌표만을 담는 역할을 수행했기 때문에, 객체 생성 시 **코드의 간결함**을 고려하여 생성자 내부에서 직접 좌표를 설정하는 방식을 선택했던 것 같습니다..  🤣

- **리팩토링 전**
`KING(new King(5, 2, PieceDirection.KING.get()));`

- **리팩토링 후**
`KING(new King(Position.of(5, 2), PieceDirection.KING.get()));`

객체를 생성자에서 직접 생성하는 방식은 클래스 간의 강한 결합이 생긴다고 생각합니다. 예를 들어 `Position` 클래스의 생성자가 변경될 경우, 이를 사용하는 모든 `Piece` 클래스에서도 변경이 발생하게 되어 코드 전반에 걸쳐 더 큰 유지보수 비용을 초래한다고 판단했습니다.

따라서 위와 같이 리팩토링으로 결합을 낮추고 향후 테스트 용이성을 향상시키기 위해 생성자에서 외부로부터 객체를 주입받는 방식이 보다 올바른 설계 방법이라고 생각합니다.

**변경 코멘트 : [refactor: 생성자 주입 형태로 변경](https://github.com/woowacourse/java-janggi/pull/43/commits/086eb9f23d89decd2476078b0912bec40a2c4f91)**

hyeonic replied:
말씀 주신 것 처럼 외부에서 생성자를 통해 주입을 받게되면 보다 더 유연한 설계가 가능하다고 생각해요. Position까지 수정하는 것을 의도한 것은 아니지만 적절히 잘 수정된 것 같습니다 👍

hyeonic commented on src/main/java/domain/piece/Position.java:
```diff
@@ -0,0 +1,64 @@
+    private final int row;
+    private final int column;
+
+    private Position(final int row, final int column) {
+        this.row = row;
+        this.column = column;
+    }
+
+    public static Position of(final int row, final int column) {
+        validateRange(row, column);
+        return new Position(row, column);
+    }
+
+    public static Position ofDirection(final int row, final int column) {
+        return new Position(row, column);
+    }
```
private을 통해 적절히 생성자를 숨기셨네요 👍

hyeonic commented on src/main/java/domain/piece/Position.java:
```diff
@@ -0,0 +1,64 @@
+        }
+    }
+
+    public Position merge(final Position other) {
+        return new Position(this.row + other.row, this.column + other.column);
+    }
```
일반적인 생성자를 활용하면 validation이 누락될 것으로 보이는데요, 아래와 같은 케이스는 어떤 결과가 나올까요?

```java
Position position = Position.of(9, 10);
Position otherPosition = Position.of(9, 10);

Position result = position.merge(otherPosition);
```

taek2222 replied:
현재 비즈니스 코드에서는 해당 예시가 발생하지 않을 것이라고 안일하게 판단하여, `Position` 클래스의 `merge` 메서드에서 검증 로직을 제외했던 것 같습니다.. 😢 개선하기 위해 여러 방향으로 리팩토링을 시도했습니다.

우선, Direction 클래스에서 기존의
`private final List<Position> direction;`
`private final List<Vector> direction;`
로 변경하였습니다.

방향을 `Position` 객체로 관리하는 것이 어색하다고 판단하여, 이동 방향의 개념을 보다 명확하게 표현할 수 있도록 `Vector` 클래스를 새롭게 정의했습니다. 🚀

또한, `Position` 클래스에서는 단순한 병합 의미를 가지던 `merge` 메서드를 제거하고, 이동의 의미를 명확히 전달할 수 있도록 `moveBy` 메서드를 새로 정의했습니다. 🚀

마지막으로, 이동 후 올바른 방향인지 검증할 수 있도록 `isMoveValid` 메서드를 추가했습니다. 이전과는 다르게 견고한 구조를 갖출 수 있도록 최대한 개선하였습니다. 🚀

**변경 점 클래스 : `Position`, `Vector`, `Direction`**

**변경 코멘트 :** 

- [refactor: record 형태 변경 및 움직임 및 검증 로직 변경](https://github.com/woowacourse/java-janggi/pull/43/commits/4e5e2b22448c40941eb57edceb481a0129924824)

- [feat: Vector 클래스 추가](https://github.com/woowacourse/java-janggi/pull/43/commits/b739d0608b47cddf3e2a3704ef002c609e331142)

- [refactor: 방향 클래스 Vector 클래스 적용](https://github.com/woowacourse/java-janggi/pull/43/commits/722e782ec44300f6fe7e7f32849baa9f0c16fcd4)

hyeonic commented on src/main/java/domain/direction/Direction.java:
```diff
@@ -0,0 +1,68 @@
+        return createPathWithoutRepeat(start);
+    }
+
+    private boolean canReachWithRepeat(final Position start, final Position target) {
+        Position result = start;
+        while (result.isValid() && !result.equals(target)) {
+            result = result.merge(direction.getFirst());
+        }
+        return result.isValid() && result.equals(target);
+    }
+
+    private boolean canReachWithoutRepeat(final Position start, final Position target) {
+        Position result = start;
+        for (Position dir : direction) {
+            result = result.merge(dir);
+        }
+        return result.equals(target);
+    }
```
result라는 표현은 너무 모호한 것 같아요. 변수의 역할이 나타날 수 있도록 적절하게 수정할 순 없을까요?

taek2222 replied:
구현 당시 시간적 압박으로 인해 네이밍을 깊게 고민하지 못하고 단순하게 지은 채 리팩토링을 놓쳤던 것 같습니다..! 🤣

이번에 좀 더 명확한 의미를 가진 이름으로 변경했습니다.

**변경 코멘트 : [refactor: 어색한 네이밍 수정](https://github.com/woowacourse/java-janggi/pull/43/commits/be141e3c016b8a26ce14da8b1f35ce535a6d5114)**

hyeonic commented on src/main/java/domain/Board.java:
```diff
@@ -0,0 +1,102 @@
+import java.util.List;
+import java.util.Map;
+
+public record Board(
```
보드에서 모든 기물에 대한 이동 전략을 관리하는 것 처럼 보이네요! 비타는 어떤 목적으로 이러한 구현을 고민하셨나요??

taek2222 replied:
실제 장기의 모습을 상상하며 설계했습니다. 🤔

**`Piece`** : 각 기물(왕, 마, 차, 포 등)은 고유한 역할과 그에 따른 이동 방향이 정해져 있다고 생각했습니다. 그래서 각 기물이 자신의 이름과 이동 방향성을 채임지고 알도록 구현했습니다. 🚀

**`Board`** : 한편, 플레이어가 직접 기물을 선택하고 움직이듯, 실제 이동의 결정은 기물이 아닌 플레이어의 책임이라고 판단하여, 페어와 논의한 결과 이를 `Board` 클래스가 담당하도록 결정했습니다. 🚀

- 각 기물이 이동할 수 있는 방향은 각 기물 스스로 책임
- 어떤 기물을 선택해 이동할지 결정하는 것은 플레이어의 역할이며, 이는 Board 클래스의 책임으로 설계

와 같은 목적으로 설계해 구현했습니다. ☺️

hyeonic commented on src/main/java/domain/piece/Pieces.java:
```diff
@@ -0,0 +1,45 @@
+
+import java.util.List;
+
+public record Pieces(List<Piece> pieces) {
```
비타는 어떤 때에 `record`를 고려하시나요? 일반적인 class와 어떠한 차이가 있을까요?

taek2222 replied:
생성자가 필드에 값을 직접 할당하는 생성자 주입 형태이거나, 불변 데이터 보장이 필요할 때는 `Record` 클래스로의 전환을 고려합니다. 🤔

이전에 `Record` 방식과 일반 `Class` 방식 중 어떤 것을 선택할지 고민한 물음에 페어와 의견이 동일할 경우나 팀에서 정한 방향대로 개선하기로 답변 받았습니다. 이에 따라 이번에 페어와 의견을 나누고 `Record` 클래스로 변환해보았습니다. 👀

hyeonic replied:
좋습니다! 그렇다면 `불변 객체`는 어떠한 이점을 가질 수 있을까요?

taek2222 replied:
불변 객체인 경우 오직 `read-only` 메서드만 제공해 데이터 변경을 제한하기 때문에 객체의 변하지 않음을 보장할 수 있습니다. 그러므로, 자연스럽게 변경 가능한 상태로 발생하는 `복잡성`을 최소화하고 `유지보수성`을 향상 시킬 수 있습니다. ☺️

또한, `버그 예방`도 할 수 있습니다. `다중 스레드` 환경에서 다양한 접근에도 상태가 변하지 않고, 변경되는 부분은 새로운 객체를 반환해 사용하기 때문에 `버그`와 `스레드 안정성`을 보장할 수 있습니다. ⚠️

hyeonic reviewed (changes requested):
안녕하세요 비타~ 장기 리뷰를 맡은 매트입니다 😃

간단히 코멘트 남겼습니다. 확인해주시고 질문이 모호하거나 이해하기 어려운 부분 있다면 언제든 편하게 DM 남겨주세요~

taek2222 committed:
refactor: 명확한 기물 클래스 네이밍 변경

taek2222 committed:
refactor: 생성자 주입 형태로 변경

- 외부의 객체를 생성해 주입을 시켜주는 형태로 변경했습니다.

taek2222 committed:
refactor: 어색한 네이밍 수정

taek2222 committed:
refactor: Position 클래스 패키지 이동

taek2222 committed:
refactor: record 형태 변경 및 움직임 및 검증 로직 변경

- 움직임을 더 명확한 네이밍으로 변경하고, Vector 클래스를 추가해 이동 파라미터를 받는 형태로 변경했습니다.

- 유효한 좌표인지는 생성자에서 검증하도록 변경하고, 이동한 후 올바른 좌표가 맞는지의 역할로 변경했습니다.

taek2222 committed:
feat: Vector 클래스 추가

- 움직이는 명확한 의미를 담기 위해 새로 정의했습니다.

taek2222 committed:
refactor: Vector, Position 클래스 변경으로 인한 수정

taek2222 committed:
refactor: 방향 클래스 Vector 클래스 적용

taek2222 committed:
test: 백터 테스트 추가

taek2222 committed:
feat: final 키워드 추가

taek2222 committed:
docs: 초기 전략 선택 기능 요구 사항 추가

taek2222 committed:
refactor: Player 필드 팀 새로 정의

taek2222 committed:
feat: 초기 전략 선택 기능 추가

- 전략 패턴 및 팩토리 패턴을 적용해 선택할 수 있도록 개선했습니다.

taek2222 committed:
refactor: Player 객체 변경점 적용

taek2222 committed:
refactor: 출력 메시지 포멧팅 정렬

taek2222 committed:
fix: 잘못된 전략 순서 변경 완료

taek2222 committed:
refactor: 비즈니스 로직 메서드 분리 적용

taek2222 commented on src/main/java/domain/direction/Direction.java:
```diff
@@ -0,0 +1,70 @@
+
+public class Direction {
+
+    private final List<Vector> direction;
+    private final boolean repeatable;
```
원래는 `Position` 이었지만, `Vector` 로 변경했습니다. 좀 더 방향성의 의미가 명확해졌나요!? 🤔

hyeonic replied:
방향을 나타내는 것 같아서 보다 더 명확해진 것 같아요!

taek2222 commented on src/main/java/domain/piece/Piece.java:
```diff
@@ -0,0 +1,55 @@
+
+    public abstract Piece updatePosition(final Position position);
+
+    public abstract boolean isKing();
+
+    public abstract boolean isCannon();
```
개선 방향을 고민하며 몇 가지 **선택지**를 고려했습니다.

1. getClass() 메서드를 활용하여 if 조건문으로 개선
2. instanceof를 활용하여 if 조건문으로 개선
3. 각 기물마다 Enum 타입을 추가하여 명시적으로 판단(필드 증가)
4. 전체적인 설계 변경
5. 기존 방식 유지

각 방법마다 장단점이 존재하지만, 비용적인 측면을 고려했을 때 현재 방식이 가장 최선이라 판단했습니다.. 🤔 이외 개선 방향이나 선택지에서의 매트 의견이 궁금합니다!

hyeonic replied:
관련 [코멘트 1](https://github.com/woowacourse/java-janggi/pull/43#r2006879468), [코멘트 2](https://github.com/woowacourse/java-janggi/pull/43#discussion_r2008666182)을 함께 다룰 수 있을 것 같아서 아래에 이어서 진행하겠습니다.

instanceof의 사용은 말씀 주신 것 처럼 캡슐화 및 추상화의 원칙을 위반할 수 있을 것 같아서 사용을 지양하는 편이 좋을 것 같아요!

먼저 비타가 설계하신 보드(현재 방식)에서는 모든 피스에 대한 이동 전략에 대해 알고 있기 때문에 이를 판단하기 위한 코드(isKing, isCannon)가 필요할 것 같습니다.

비타가 위에서 설명 주신 것 처럼 실제 장기 모습을 기반으로 플레이어가 보드의 모든 기물을 움직일 수 있도록 하다보면 결국 기물을 움직일 때 마다 기물의 내부 필드 정보를 꺼내 보드 내에서 모든 것을 판단해야 할 것 같아요. 각각의 정보를 가진 전문가들에게 메시지를 던지는 형태로 개선할 순 없을까요? 현재는 궁, 궁성을 제외하고 있으며 king, guard를 제외한 일반적인 기물의 이동 전략만 고려하고 있는데요, 만약 기물의 추가적인 이동 전략을 고려할 땐 어떻게 될까요? 보드의 책임이 너무 비대해지진 않을까요? 

요 부분은 사람마다 요구사항을 어떻게 해석하고 설계하는지의 관점에 따라 달라질 수 있어서 간단히 제 의견 남겨보았어요!

taek2222 replied:
설계 방향에 대한 꼼꼼한 피드백 정말 감사합니다! ☺️

말씀해주신 부분은 초반 설계 당시 많이 고민했던 지점이었습니다. 📍

특히 각 기물이 자신의 이동 방향은 알고 있지만, 실제로 이동 가능 여부를 판단하는 로직을 기물 자체로 옮기는 부분에서 고민이 깊었습니다.

이렇게 설계하면 결국 보드가 가진 전체 기물의 상태를 각 기물이 알아야 하는 상황이 생겼고, 이는 오히려 기물이 지나치게 많은 정보를 가지게 되는 문제로 이어진다고 판단했습니다. 🤔

매트께서 주신 의견처럼 현재 보드의 책임이 다소 과중하다고 생각하고 있어서, 이 부분은 추가적인 피드백을 반영하여 계속 개선해 나가겠습니다. 

이번 PR 내에서 마지막으로 고민해 보고, 개선된 부분이 있다면 코멘트를 통해 별도로 정리하여 공유하겠습니다.

다시 한번 좋은 의견 주셔서 감사드립니다! 🙇🏻

taek2222 replied:
중간 기물의 개수 별로 검증할 수 있는 책임을 Pieces 스스로가 할 수 있도록 변경해보았습니다. ☺️

**변경 코멘트 :**
- [refactor: 중간 기물 검증 스스로 하도록 개선](https://github.com/woowacourse/java-janggi/pull/43/commits/d654e0bb5636fb134a0121396dafde1ff4a24c56)
- [refactor: 중간 기물 검증은 Piece 책임으로 변경](https://github.com/woowacourse/java-janggi/pull/43/commits/ef5037102e29feac35dc3505292ba45692690653)

위 변경은 어느정도 적절했다고 생각합니다. 하지만, 아래 부분이 상당히 고민이 깊어서 새로운 방법을 시도했습니다.

- 도착지에 우리팀 기물 존재 검증
- 포인 경우 우리팀, 상대팀 기물의 추가적인 여러 검증

위 검증과 관련된 책임은 pieces의 존재를 알고 있는 board가 담당하는 것이 적절하다고 판단했습니다. 이에 따라, 과도한 board의 책임을 분리하기 위해 별도의 행위 객체를 생성하고 로직을 분리하는 방향으로 개선해보았습니다. 🚀

그러나 결과적으로 단순히 메서드만 분리된 느낌이 강했고, 기대했던 만큼의 효과를 얻지는 못했습니다. 오히려 제3자가 보기에 로직이 더 이해하기 어려워졌을 수도 있다는 생각이 듭니다. 🤔

이 점에 대해 매트의 의견이 궁금합니다. 만약 개선이 아니라 복잡성을 증가시킨 것이라면, 기존 방식으로 롤백하는 것도 고려하고 있습니다. 🔥

**변경 코멘트 :**
- [refactor: 검증 및 움직이는 담당하는 객체 분리](https://github.com/woowacourse/java-janggi/pull/43/commits/9b0150a2a4316da7812f9956195f8dbe8893a4d7) (... 여러 코멘트 중 가장 최종 코멘트입니다.)

hyeonic replied:
위와 같은 이유로 `PieceMover`가 탄생하게 되었군요. 메서드 분리 관점에서는 좋지만 말씀 주신 것 처럼 흐름을 따라가기 어려울 것 같기도 합니다. 

결국 이동 전략을 각 기물의 특성에 기대는 것이 아니라 보드에 존재하던 이동 전략을 단순히 메서드를 분리하는 것 처럼 보입니다. 요부분은 DM으로 질문 주신 부분 추가적으로 이야기 나눠보고 나온 결론으로 개선해보면 좋을 것 같아요.

taek2222 committed:
refactor: PieceInit 클래스 및 테스트 삭제

taek2222 commented on src/main/java/domain/spatial/Position.java:
```diff
@@ -0,0 +1,35 @@
+        return vector.applyTo(this);
+    }
+
+    public boolean isMoveValid(final Vector vector) {
```
움직일 수 있는지 판단을 다음과 같이 에러 발생 여부로 판단했습니다..! 혹시 이러한 설계도 가능한가요? 🤔

hyeonic replied:
충분히 활용은 가능하겠지만 vertor에게 boolean 타입을 반환하는 메서드를 구성하는 것이 좀 더 자연스러워보이네요!

taek2222 replied:
`Vector` 클래스의 책임으로 두려고 했지만, 좌표의 상수를 `Position` 클래스에 정의되어 있어 이와 같이 수정했습니다. 🚀

**변경 코멘트 :**
[refactor: 이동 가능 판정 로직 개선](https://github.com/woowacourse/java-janggi/pull/43/commits/7baafb132b4f72ffb442d899e5b7ff051b230022)

taek2222 commented on src/main/java/domain/spatial/Position.java:
```diff
@@ -0,0 +1,35 @@
+        validateRange(row, column);
+    }
+
+    public Position moveBy(final Vector vector) {
+        return vector.applyTo(this);
+    }
```
`Position` 클래스와 `Vector` 클래스를 분리하면서, 이동을 담당할 객체를 어디에 두는 것이 적절한지 고민했습니다.

`Position`은 고정된 하나의 좌표를 의미하고,
`Vector`는 이동할 방향을 나타내므로,
`Vector` 클래스에서 이동 후의 `Position` 객체를 생성하는 방식으로 설계했습니다.

다만, 두 객체가 긴밀하게 연결되어 있음에도 불구하고, `Vector`에서 `Position`을 생성하는 방식이 다소 어색하게 느껴집니다.

매트가 보기에는 이 방식이 적절한지, 혹은 더 나은 방향이 있을지 궁금합니다! 😊

hyeonic replied:
말씀 주신 것 처럼 결국 `상호 참조`가 발생할 수 있을 것 같아요. Vector는 말 그대로 방향에 대한 정보만 가진 VO로 활용되고 Position을 생성하는 것은 Position 객체 내에서 소화하는 것도 좋아보이네요!

```java
    public Position moveBy(final Vector vector) {
        return new Position(this.row() + vector.moveRow(), this.column() + vector.moveColumn());
    }
```

taek2222 replied:
의견 꼼꼼하게 봐주셔서 감사합니다! ☺️

한 번 새로운 도전을 해보고 싶었고, 의미를 확실하게 해보고 싶었습니다. 상호 참조의 문제가 더 단점이 강하다고 느꼈기 때문에 아래와 같이 개선했습니다! 🙇🏻

**변경 코멘트 :**
[refactor: Vector, Position 상호 참조 문제 개선](https://github.com/woowacourse/java-janggi/pull/43/commits/47a5c532c4d06d0950e1374b080a1adda4fff27c)

taek2222 reviewed (commented):
소중한 시간 내어 리뷰해 주셔서 감사합니다!

시간이 다소 걸리긴 했지만, 고민할 부분이 있어서.. 하하 😂

좋은 하루 보내세요! 🙌

hyeonic commented on src/main/java/domain/strategy/SettingUpStrategy.java:
```diff
@@ -0,0 +1,8 @@
+import domain.Player;
+import domain.piece.Pieces;
+
+public interface SettingUpStrategy {
+    Pieces initPieces(final Player player);
+}
```
[코멘트](https://github.com/woowacourse/java-janggi/pull/43#discussion_r2009000854) 답변 아래에 추가적으로 남길게요.

초기화 세팅 전략을 인터페이스로 잘 추상화하였네요 😃

여담으로 `Strategy`같이 특정 패턴을 뜻하는 이름을 클래스명으로 활용하지 않아도 괜찮을 것 같아요. 가령 `settingUpInitializer`와 같은 이름은 보다 더 초기 세팅을 위한 목적의 인터페이스임을 드러낼 수도 있지 않을까 합니다 ㅎㅎ

taek2222 replied:
오호! 사실 인터페이스와 상속 적용이 아직 생소해서 특정 패턴을 클래스명으로 활용하지 않는다는 점을 몰랐습니다. 😮 새로운 사실을 알게 되어 정말 감사합니다!

**변경 코멘트 :**
[refactor: 이동 전략 인터페이스 및 구현체 네이밍 변경](https://github.com/woowacourse/java-janggi/pull/43/commits/3d33fc758603c0ff3a6f5ec1de02656b2b3231c2)

hyeonic replied:
활용되지 않는다기보단 변수명에 타입 정보를 기재하는 것을 지양하는 것과 비슷한 맥락으로 말씀드렸어요! 특정 패턴에 대한 의도가 클래스명으로 활용되면 추후 다른 패턴은 고려하기 어렵지 않을까 생각이 되어서요.

taek2222 replied:
> 특정 패턴에 대한 의도가 클래스명으로 활용되면 추후 다른 패턴은 고려하기 어렵지 않을까 생각이 되어서요.

혹시 말씀하신 것이, 하나의 패턴으로 설계된 구조에서 다른 패턴으로 변경되는 상황을 의미하는 걸까요?

만약 그렇다면, 설계된 패턴이 다른 패턴으로 바뀌는 일이 자주 발생하나요? (아직 그런 상황을 경험하지 못해서인지.. 잘 이해가 가지 않아서.. 🤔)

hyeonic commented on src/main/java/domain/strategy/InnerElephantStrategy.java:
```diff
@@ -0,0 +1,83 @@
+
+public class InnerElephantStrategy implements SettingUpStrategy {
+
+    @Override
+    public Pieces initPieces(final Player player) {
+        if (player.team().equals(Team.HAN)) {
+            return initHanPieces();
+        }
+        return initChoPieces();
+    }
```
player에게 같은 팀인지 메시지를 던지는 형태로 개선해보는 건 어떨까요??

taek2222 replied:
헉,, 꼼꼼하게 봐주셔서 감사합니다. ☺️

**변경 코멘트 :**
- [feat: Han 팀 확인 기능 추가](https://github.com/woowacourse/java-janggi/pull/43/commits/66f082f661a26ff4a6a15b770df26fcb095ed5a6)

- [refactor: Player 기능의 메시지 활용 로직 개선](https://github.com/woowacourse/java-janggi/pull/43/commits/9c3f3cc6d049e6ca84a128b4757f8e9b104e48bd)

hyeonic commented on src/main/java/domain/piece/Pieces.java:
```diff
@@ -0,0 +1,46 @@
+                .orElseThrow(IllegalArgumentException::new);
+    }
+
+    public int countPiecesInPositions(final List<Position> positions) {
+        return (int) positions.stream()
+                .filter(position ->
+                        pieces.stream().anyMatch(piece -> piece.isSamePosition(position)))
+                .count();
+    }
```
stream을 적극적으로 활용하고 계신데요! java의 Stream은 어떠한 특성과 이점을 가지고 있나요??

taek2222 replied:
현재 코드를 기준으로 보면, 가장 큰 이점은 가독성이 확실히 향상되었다는 점입니다. 기존 for문을 예시로 보았을 때, Depth가 깊어지고 count 변수를 추가로 생성해야 하는 문제가 있었습니다.

혹시 제가 놓치고 있는 부분이 있을까요? Stream API를 습관적으로 사용하다 보니 치명적인 문제를 간과하고 있는 것은 아닌지 걱정됩니다. 🫢

``` java
    public int countPiecesInPositions(final List<Position> positions) {
        return (int) positions.stream()
                .filter(position ->
                        pieces.stream().anyMatch(piece -> piece.isSamePosition(position)))
                .count();
    }

    public int countPiecesInPositionsEx(final List<Position> positions) {
        int count = 0;
        for (Position position : positions) {
            for (Piece piece : pieces) {
                if (piece.isSamePosition(position)) {
                    count++;
                }
            }
        }
        return count;
    }
```

hyeonic replied:
가독성 측면의 이점도 있지만 제가 생각했을 때 stream의 가장 큰 특성은 중간/종단 연산과 lazy evaluation 등이 있을 것 같아요. 여유가 있다면 관련 키워드 학습해보면 좋을 것 같습니다 ㅎㅎ

hyeonic commented on src/main/java/domain/piece/category/Cannon.java:
```diff
@@ -0,0 +1,33 @@
+
+public class Cannon extends Piece {
+
+    private static final String NAME = "c";
```
이러한 네임 정보는 View에서만 활용되는 정보라 도메인 내에서 View에 의존적이라 생각이 들기도 하네요!

taek2222 replied:
짚어주신 부분 정말 감사합니다! ☺️
충분히 고민한 끝에 개선 방향을 선택했지만, ‘이게 맞는 걸까?’ 하는 고민이 남아 다시 질문드리고 싶습니다. 🤔

`Domain`은 `View`를 몰라야 하지만, `View`는 `Domain`을 알아도 된다고 생각합니다. 그래서 `PieceView` 상수를 등록하고, 각 클래스에서 이름을 출력할 수 있도록 설계했습니다.

다만, `PieceView` 클래스 내부에서도 `.getClass()`를 활용한` 리플렉션`으로 비교를 수행하는데, `View`에서 사용하는 것이므로 괜찮지 않을까? 하는 생각이 들었습니다. 🤔 (`View` 단위는 FE의 관점의 단위라고 생각해, `Domain`의 분리를 우선 시 하는 방향으로.. 맞게 판단한 걸까요? 😐)

매트가 바라보는 시선에서는 이 설계에 대해 어떻게 생각하시나요?

**변경 코멘트 :**
- [refactor: 기물마다 이름 지정 리팩토링](https://github.com/woowacourse/java-janggi/pull/43/commits/0da066a75c33a98b68b362faac79f5faa3d70133)

- [refactor: 사용하지 않은 기물의 이름 상수 삭제](https://github.com/woowacourse/java-janggi/pull/43/commits/a5fd5a8c173c793ac1425093da48fd6ec4ea2d2e)

hyeonic replied:
> Domain은 View를 몰라야 하지만, View는 Domain을 알아도 된다고 생각합니다. 

위 의견에 동의합니다! 말씀 주신 것 처럼 class 정보 외에는 해당 객체를 특정하기 어렵긴 할 것 같네요.

hyeonic reviewed (changes requested):
안녕하세요 비타! 남겨주신 부분 답변 및 추가 질문 남겼습니다 ㅎㅎ 혹여나 제가 놓친 질문이 있거나 이해가 어려운 부분이 있다면 언제든 편하게 DM 남겨주세요!

taek2222 committed:
refactor: Vector, Position 상호 참조 문제 개선

- Vector 클래스 Record 형식으로 변경했습니다.

taek2222 committed:
refactor: Vector 클래스 이동 메서드 삭제

taek2222 committed:
refactor: 이동 가능 판정 로직 개선

taek2222 committed:
refactor: 이동 전략 인터페이스 및 구현체 네이밍 변경

taek2222 committed:
refactor: 이동 전략 추상 클래스로 변경 및 중복 로직 제거

taek2222 committed:
feat: Han 팀 확인 기능 추가

taek2222 committed:
refactor: Player 기능의 메시지 활용 로직 개선

taek2222 committed:
refactor: 기물마다 이름 지정 리팩토링

taek2222 committed:
refactor: 사용하지 않은 기물의 이름 상수 삭제

taek2222 committed:
refactor: 중간 기물 검증 스스로 하도록 개선

taek2222 committed:
refactor: 중간 기물 검증은 Piece 책임으로 변경

taek2222 committed:
refactor: 도착지 기물 삭제 로직 객체 분리

taek2222 committed:
refactor: 기물 검증 과정 분리

taek2222 committed:
refactor: 도착지 기물 위치 삭제 pieces 책임 변경

taek2222 committed:
refactor: Board 메서드 구체적인 명칭으로 변경

taek2222 committed:
refactor: 검증 및 움직이는 담당하는 객체 분리

hyeonic reviewed (changes requested):
안녕하세요 비타! coderabbitai가 어떤 기준으로 추가 코멘트를 남기는지 정확히 확인은 안되었는데 우선 배제하고 간단히 의견 남겼습니다. DM으로 전체적인 구조 관련하여 추가 질문 주셔서 관련하여 같이 논의해보고 이후 개선 진행해보면 좋을 것 같아요!

taek2222 committed:
test: 경로에 배치된 기물이 없는지 판단하는 테스트 작성

taek2222 committed:
feat: 경로의 피스가 있는 지 검사하는 기능 추가

taek2222 committed:
test: 경로의 배치된 기물이 없는 지 판단 테스트 케이스 추가

taek2222 committed:
test: 이동 기물과 도착지 기물이 같은 지 판단 테스트 작성

taek2222 committed:
test: 기물 정보 리스트 테스트로 변경

taek2222 committed:
test: 기물이 존재하는 지 판단하는 테스트 작성

taek2222 committed:
feat: 기물이 존재하는 지 판단하는 기능 추가

taek2222 committed:
test: 기물 존재 테스트 파라미터 추가

taek2222 committed:
test: 경로에 기물의 갯수 반환 테스트로 변경

taek2222 committed:
feat: 경로에 기물의 개수를 반환하는 기능 추가

taek2222 committed:
test: 같은 기물인지 판단하는 테스트 작성

taek2222 committed:
feat: 같은 기물인지 판단하는 기능 추가

taek2222 committed:
refactor: 가독성 향상을 위한 메서드 네이밍 변경

taek2222 committed:
test: 이동 기물과 도착지 기물이 같은지 판단 테스트 작성

taek2222 committed:
feat: 이동 기물과 도착지 기물이 같은지 판단 기능 추가

taek2222 committed:
fix: 마지막 경로 버그 수정

taek2222 committed:
test: 가독성 향상을 위한MethodSource 테스트 CsvSource 형태로 변경

taek2222 committed:
test: 이동 기물과 같은 기물이 경로에 있는지 테스트 작성

taek2222 committed:
feat: 이동 기물과 같은 기물이 경로에 있는지 판단 기능 추가

taek2222 committed:
refactor: 사용하지 않은 상태 삭제

taek2222 committed:
test: 각 기물의 움직임 검증 테스트 작성

taek2222 committed:
feat: 각 기물의 움직임 검증 추가 (포 제외)

taek2222 committed:
test: 포의 이동 경로를 검증하는 테스트 작성

taek2222 committed:
feat: 포의 이동 경로 검증하는 기능 추가

taek2222 committed:
refactor: 구조 변경으로, Piece 메서드 정리

taek2222 committed:
refactor: 구조 변경으로, 메서드 추가 및 삭제

taek2222 committed:
refactor: 이동 검증 구조 변경

taek2222 committed:
refactor: 구조 변경으로, Piece 메서드 정리

taek2222 committed:
fix: 이동 가능 판정 과정의 오류 수정

taek2222 committed:
refactor: 한 칸 이동은 검증이 필요 없어 삭제

taek2222 committed:
test: 경로 검증 로직 변경으로, 테스트 변동 수정

taek2222 committed:
test: 여러 구조 변경으로 테스트 수정

taek2222 committed:
refactor: 최종 목적지까지 반환하도록 수정

taek2222 committed:
refactor: 중간 경로만 카운트 하도록 변경

taek2222 committed:
refactor: 전체적인 로직 메서드 분리 및 정리

taek2222 committed:
refactor: 에러 메시지 출력 및 접두사 제거

taek2222 committed:
refactor: 움직임 객체 검증 로직 가독성 개선

taek2222 commented:
@coderabbitai full review

taek2222 commented on src/main/java/domain/strategy/SettingUpInitializer.java:
```diff
@@ -0,0 +1,65 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class SettingUpInitializer {
```
`SettingUpInitializer`의 초기화 로직에서 중복을 줄이기 위해 추상 클래스로 변경했습니다. 이와 같은 방식으로 개선했는데, 매트의 의견이 궁금합니다. 🤔

또한, `SettingUpInitializer`를 `abstract class`로 선언하면서 중복을 최소화하기 위해 접근 제어자를 명시하지 않고 `package-private` 형태로 설정했습니다. 현재 같은 패키지 내에서는 접근이 가능하지만, 패키지를 이동하면 하위 클래스라도 접근할 수 없어 컴파일 에러가 발생합니다. ⚠️

이런 상황을 고려할 때, 미리 `protected` 접근 제어자를 사용하는 것이 좋을까요, 아니면 현재처럼 `package-private` 형태로 두는 것이 더 적절할까요?

hyeonic replied:
protected가 좀 더 명시적으로 느껴지긴 합니다. 상속을 바라보는 입장에서 하위 타입만 사용 하도록 제한 할 수 있으니깐요 ㅎㅎ

taek2222 replied:
아래와 같이 변경했습니다. 다만 상속 개념에 아직 익숙하지 않아서 그런지, 현재처럼 부모 클래스의 중복을 개선한 로직이 담긴 메서드를 자식 클래스에서 호출하고, 이를 활용해 `Pieces`를 생성해 반환하는 흐름이 적절한지 조금 고민이 되네요. 🤔

**변경 코멘트 :**
[refactor: 상속 접근 제어자 정의](https://github.com/taek2222/java-janggi/commit/efe8b73fb0fede0253c6d3253d2a71ffc3a5d70a)

hyeonic replied:
왜 어색하다고 생각하시나요? 결국 자식 객체는 부모 객체의 의존성을 강하게 가져가기 때문에 그 부분이 어색하다면 상속이 고려하는 것이 적절한지 한 번 고민해봐도 좋을 것 같아요! 지금 로직을 수정하라는 것은 아니니 참고만 해주세요.

taek2222 commented on src/main/java/domain/piece/category/Horse.java:
```diff
@@ -0,0 +1,33 @@
+import domain.piece.Piece;
+import domain.spatial.Position;
+
+public class Horse extends Piece {
```
만약, 바보 같은.. 질문일 수 있지만.. 🤣 부모 클래스가 필드 2개를 가지고 있고, 자식 클래스가 필드 1개를 추가로 가진 경우,

> 7️⃣ **3개 이상의 인스턴스 변수를 가진 클래스를 만들지 않는다.**
> - 너무 많은 상태를 가지지 않도록 조심하라.

이 원칙을 위반하는 것에 해당하나요..? 🤔

hyeonic replied:
구체적인 타입인 Horse를 사용할 경우 3개 이상의 필드를 포함한다고 생각해야 할 것 같습니다. 왜 너무 많은 상태를 가지지 않도록 해야 할까요?

taek2222 replied:
정말 단순하게 생각해보면, 객체가 많은 상태를 가지게 될수록 그에 따라 행동도 자연스럽게 증가하고, 결국 객체가 담당하는 명확한 책임을 파악하기 어려워지는 것 같습니다. 🤔

또한 테스트 코드는 하나의 문서처럼 중요하다고 생각하는데, 상태가 많아질수록 테스트 케이스도 많아져서 오히려 그 테스트들이 방대해지고, 파악이 어려운 문서가 될 수 있을 것 같습니다.

결론적으로, 상태가 많아지면 복잡도가 급격히 증가하고, 여러 측면에서 관리나 유지보수에 큰 단점이 생길 수 있다고 생각합니다. 👋

hyeonic replied:
좋습니다. 상태가 많아지면 해당 객체의 책임이 막중한 게 아닌지 고민해보면 좋을 것 같아요. 단순히 `만들지 마라!`에서 멈추기 보다 `왜 못 만들게 하는거지?`를 고민하는게 미션을 통해 성장할 수 있는 포인트가 더 많지 않을까 생각이 됩니다 😃

taek2222 reviewed (commented):
소중한 시간 내어 리뷰해 주셔서 감사합니다. ☺️

오전에 논의한 변경점을 반영하여 전체적으로 로직을 수정했습니다. 각 기물이 스스로 검증 책임을 가지도록 수정했으며, 이에 따라 이동 과정의 검증을 담당하는 `MoveInfos`, `MoveInfo` 클래스를 새롭게 추가했습니다. 개선된 방향성에 대해 의견 주시면 감사하겠습니다.

또한, 이번 수정에 집중하기 위해 일부 테스트 코드가 삭제되었습니다. 추후 테스트 코드를 보완할 예정이며, 현재는 변경된 방향성과 비즈니스 로직을 중심으로 피드백을 주시면 감사하겠습니다. 🙇🏻

hyeonic commented on src/main/java/domain/piece/category/Cannon.java:
```diff
@@ -0,0 +1,41 @@
+        return CATEGORY;
+    }
+
+    @Override
+    public Cannon move(final Position position, final MoveInfos moveInfos) {
+        validateMove(moveInfos);
+        return new Cannon(position, directions);
+    }
```
현재 구조는 플레이어가 잘못 이동하면 즉시 게임이 종료될 것으로 보여요. 그것을 의도하신 건가요??

![image](https://github.com/user-attachments/assets/c85ff11c-6e25-47ba-b045-50d1842e6efd)

taek2222 replied:
의도한 부분은 아닙니다! 🤣 
장기 게임중 기물 한 번 잘못 움직여서 게임이 끝나면.. 말이 안되긴합니다. 하하 😝

테스트 코드를 작성 때문인지, 직접 실행해서 확인하는 과정을 거의 거치지 않다 보니 놓친 부분이 있었던 것 같습니다.

**변경 코멘트 :**
[feat: 예외 발생 상황도 재입력 기능 추가](https://github.com/taek2222/java-janggi/commit/72c9b7c42a44512594f834a6aab1b57b87d11f8e)

hyeonic commented on src/main/java/domain/piece/category/Cannon.java:
```diff
@@ -0,0 +1,41 @@
+        return new Cannon(position, directions);
+    }
+
+    private void validateMove(final MoveInfos moveInfos) {
+        if (moveInfos.countPiecesInIntermediatePath() != PIECES_TO_PASS) {
+            throw new IllegalArgumentException("포는 중간에 기물이 " + PIECES_TO_PASS + "개여야 합니다.");
+        }
+
+        if (moveInfos.isSameAsTargetPiece(CATEGORY)) {
+            throw new IllegalArgumentException("포는 상대 포를 잡을 수 없습니다.");
+        }
+
+        if (moveInfos.hasSamePieceCategoryInPath(CATEGORY)) {
+            throw new IllegalArgumentException("포는 다른 포를 지나칠 수 없습니다.");
+        }
+    }
```
현재 구조도 충분하지만 좀 더 수정할 여지가 있다면 포의 이동 전략을 3개로 나눠서 N개의 전략을 기반으로 검증할 수도 있을 것 같아요.
1. 중간 기물 체크
2. 목적지 기물 체크
3. 중간 기물이 존재할 때 포인지 체크

이건 참고만 해주세요!

taek2222 replied:
매트의 피드백에서 정확히 어떤 방향을 제안하신 건지는 파악하기.. 힘들었습니다. 🤔

두 가지 의미가 있었습니다.

1. 메시지 기반의 현재 코드만으로는 각각의 검증이 어떤 의미인지 명확히 드러나지 않는다는 점에서, 검증 로직을 메서드로 분리해 의미를 드러내는 방향

2. 검증을 각각의 클래스로 추상화해 `List<CannonMoveValidation>` 형태로 구성함으로써 새로운 검증이 추가될 때에도 기존 코드를 수정하지 않도록 하는, `개방 폐쇄 원칙`을 적용한 전략 패턴 방식 제안

우선 저는 첫 번째 맥락으로 이해하고, 메서드를 분리하여 리팩토링을 진행했습니다. 🚀

**변경 코멘트 :**
- [refactor: 예외 검증 메서드 분리](https://github.com/taek2222/java-janggi/commit/b6ccffeabfd3bbde8f609a561d7510f728f6a4b9)

hyeonic replied:
제가 너무 모호하게 답변 드린 것 같네요!

저는 2번 의미로 말씀 드린 거예요. 이동 전략을 좀 더 세분화하여 관리하면 더 작은 단위로 움직임을 제어할 수 있지 않을까 생각이 들어서요. 이부분은 그냥 제 의견이라 참고만 하시라고 남겨보았어요.

hyeonic commented on src/main/java/domain/MoveInfos.java:
```diff
@@ -0,0 +1,30 @@
+        this.moveInfos = moveInfos;
+    }
+
+    public int countPiecesInIntermediatePath() {
+        List<MoveInfo> intermediatePaths = moveInfos.subList(0, moveInfos.size() - 1);
+        return (int) intermediatePaths.stream()
+                .filter(MoveInfo::hasPieceInPath)
+                .count();
+    }
```
coderabbit이 남겨준 것 처럼 subList 전에 길이가 충분한지 체크해주면 좋을 것 같아요!

taek2222 replied:
모든 기물이 한 번씩은 움직인다고 생각해 이와 같이 구현했습니다! 🤔 하지만, 장기 룰에

> 궁은 손 위에서 한번 회전시킨 후 제자리에 두는 것으로 차례를 쉴 수 있다

내용이 존재하는 점을 파악했습니다..! 감사합니다!

**변경 코멘트:**
- [feat: 길이 충분 검증 로직 추가](https://github.com/taek2222/java-janggi/commit/70eda327a83281d8d66835a79fc81a4d4d4e4291)

hyeonic commented on src/main/java/domain/MoveInfos.java:
```diff
@@ -0,0 +1,30 @@
+import domain.piece.category.PieceCategory;
+import java.util.List;
+
+public class MoveInfos {
```
해당 정보를 통해 기물 내에서 직접 이동 여부를 판단할 수 있을 것 같아요 👍 잘 개선된 것 같습니다 ㅎㅎ

hyeonic commented on src/main/java/domain/Board.java:
```diff
@@ -0,0 +1,71 @@
+) {
+
+    public void moveAndCapture(final Player player, final Position startPosition, final Position targetPosition) {
+        Pieces playerPieces = playerPiecesMap.get(player);
+        Pieces opponentPieces = getOppositePieces(player);
```
타입 정보를 필드나 변수명으로 사용하는 것은 지양하는 편이 좋아보여요.

taek2222 replied:
놓치는 요소 중 하나였습니다. 꼼꼼하게 봐주셔서 감사합니다. ☺️ 계속 리팩토링 하면서 추가 개선하겠습니다.

**변경 코멘트:**
- [refactor: 필드나 변수명에 필드명 사용 지양 개선](https://github.com/taek2222/java-janggi/commit/8fbb0fbd40e669ec0af84408d2e0263cbaf39379)

hyeonic reviewed (approved):
안녕하세요 비타! 대부분 요구사항 잘 반영해주셔서 이만 머지해도 좋을 것 같아요 ㅎㅎ 이후 단계 진행하면서 더 많은 이야기 나눠보면 좋을 것 같습니다!
